runtime/near-vm-runner/src/preload.rs

```diff
@@ -70,46 +132,83 @@ impl ContractCaller {
 
     pub fn run_preloaded<'a>(
         self: &mut ContractCaller,
-        prepared: &ContractCallPrepareResult,
+        preloaded: &ContractCallPrepareResult,
         method_name: &str,
         ext: &mut dyn External,
         context: VMContext,
-        vm_config: &'a VMConfig,
         fees_config: &'a RuntimeFeesConfig,
         promise_results: &'a [PromiseResult],
         current_protocol_version: ProtocolVersion,
         profile: ProfileData,
     ) -> (Option<VMOutcome>, Option<VMError>) {
-        match self.prepared.get(prepared.handle) {
+        match self.preloaded.get(preloaded.handle) {
             Some(call) => {
                 let call_data = call.rx.recv().unwrap();
                 match call_data.result {
                     Err(err) => (None, Some(err)),
-                    Ok(module) => match module {
-                        Wasmer0(module) => run_wasmer0_module(
-                            module,
-                            method_name,
-                            ext,
-                            context,
-                            vm_config,
-                            fees_config,
-                            promise_results,
-                            profile,
-                            current_protocol_version,
-                        ),
-                        Wasmer1(module, store) => run_wasmer1_module(
-                            &module,
-                            &store,
-                            method_name,
-                            ext,
-                            context,
-                            vm_config,
-                            fees_config,
-                            promise_results,
-                            profile,
-                            current_protocol_version,
-                        ),
-                    },
+                    Ok(module) => {
+                        match (&module, &mut self.vm_data_private, &self.vm_data_shared) {
+                            (
+                                Wasmer0(module),
+                                VMDataPrivate::Wasmer0(memory),
+                                VMDataShared::Wasmer0,
+                            ) => {
+                                let mut new_memory;
+                                run_wasmer0_module(
+                                    module.clone(),
+                                    if SHARE_MEMORY_INSTANCE {
+                                        memory.as_mut().unwrap()
+                                    } else {
+                                        new_memory = WasmerMemory::new(
+                                            self.vm_config.limit_config.initial_memory_pages,
+                                            self.vm_config.limit_config.max_memory_pages,
+                                        )
+                                        .unwrap();
+                                        &mut new_memory
+                                    },
```

Comment:
It feels like we do a the same check twice here: `SHARE_MEMORY_INSTANCE` is true iff `memory.as_mut()` is not None. So perhaps use this instead?

```suggestion
                                    module.clone(),
                                    memory.as_mut().unwrap_or_else(|| {
                                        new_memory = WasmerMemory::new(
                                            self.vm_config.limit_config.initial_memory_pages,
                                            self.vm_config.limit_config.max_memory_pages,
                                        )
                                        .unwrap();
                                        &mut new_memory
                                    }),
```

Reply:
Indeed, somewhat reworked this place.